### PR TITLE
 Single function hostdev hotplug

### PR DESCRIPTION
**Pull request by prudmirya**
This Test Case will do hotplug the single function of NIC or Infinibd adapter then it will ping to ouside world from hotplugged interface, guest life cylce, flood ping, virsh dump

then it will hotunplug the device.

Signed-off-by: prudhvi <mprudhvi@linux.vnet.ibm.com>

### CODE REVIEW

**Comment by balamuruhans** [comment]
current commit msg doesn't have header and description appropriately, please change it 

1. header/subject/Title (50 characters)
2. proper description (explain the need for the change)
3. signed-off-by:

File: libvirt/tests/src/libvirt_pci_passthrough_hotplug.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+            test.fail("passthroughed adapter not found after suspend/resume")
+        # Reboot(not supported on RHEL6)
+        result = virsh.reboot(vm_name, ignore_status=True, debug=True)
+        supported_err = 'not supported by the connection driver:\
+                         virDomainReboot'
+        if supported_err in result.stderr.strip():
+            logging.info("Reboot is not supported")
+        else:
+            libvirt.check_exit_status(result)
+        session = vm.wait_for_login()
```
Comment (balamuruhans) [inline libvirt/tests/src/libvirt_pci_passthrough_hotplug.py]: session is not used in this function, vm.wait_for_login() is sufficient

File: libvirt/tests/src/libvirt_pci_passthrough_hotplug.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+    # Check the parameters from configuration file.
+    for each_param in params.itervalues():
+        if "ENTER_YOUR" in each_param:
+            test.cancel("Please enter the configuration details of %s."
+                        % each_param)
+    vmxml = VMXML.new_from_inactive_dumpxml(vm_name)
+    backup_xml = vmxml.copy()
+    devices = vmxml.get_devices()
+    nic_id = None
+    pci_value = None
```
Comment (balamuruhans) [inline libvirt/tests/src/libvirt_pci_passthrough_hotplug.py]: nic_id and pci_value can be removed

File: libvirt/tests/src/libvirt_pci_passthrough_hotplug.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+                    test.fail("Hostdev node reattach failed")
+                driver_name = process.system_output(cmd, shell=True).strip()
+                if driver_name == "vfio-pci":
+                    test.error("driver bind failed after reattach")
+
+    def check_attach_pci():
+        session = vm.wait_for_login()
+        output = session.cmd_output("ip link")
+        nic_list_after = str(output.splitlines())
+        if nic_list_after != nic_list_before:
```
Comment (balamuruhans) [inline libvirt/tests/src/libvirt_pci_passthrough_hotplug.py]: `return nic_list_after != nic_list_before`

File: libvirt/tests/src/libvirt_pci_passthrough_hotplug.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+    dev.hostdev_type = 'pci'
+    dev.managed = 'no'
+    dev.source = dev.new_source(**pci_address)
+    dev = str(dev)
+    temp, dev = dev.split("?>", 1)
+    fp.write(dev)
+    fp.close()
+
+    def detach_device(pci_devs, pci_ids):
+        # detaching the device from host
```
Comment (balamuruhans) [inline libvirt/tests/src/libvirt_pci_passthrough_hotplug.py]: please have docstrings

File: libvirt/tests/src/libvirt_pci_passthrough_hotplug.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+                logging.debug("device alreay detached")
+            else:
+                if virsh.nodedev_detach(pci_value).exit_status:
+                    test.error("Hostdev node detach failed")
+                driver_name = process.system_output(cmd, shell=True).strip()
+                if driver_name != "vfio-pci":
+                    test.error("driver bind failed after detach")
+
+    def reattach_device(pci_devs, pci_ids):
+        # reattach the device to host
```
Comment (balamuruhans) [inline libvirt/tests/src/libvirt_pci_passthrough_hotplug.py]: docstrings for here and elsewhere for other methods

File: libvirt/tests/src/libvirt_pci_passthrough_hotplug.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+from virttest import data_dir
+from virttest.libvirt_xml.devices.controller import Controller
+from virttest import utils_package
+from provider import libvirt_version
+from virttest.utils_test import libvirt
+
+
+def run(test, params, env):
+    """
+    Test for PCI single function device(NIC or Infiniband)
```
Comment (balamuruhans) [inline libvirt/tests/src/libvirt_pci_passthrough_hotplug.py]: the params you have introduced in cfg for NIC will be same for infiniband too?

File: libvirt/tests/cfg/libvirt_pci_passthrough_hotplug.cfg [hunk 7]
```diff
@@ -0,0 +1,29 @@
+- libvirt_pci_passthrough_hotplug:
+    type = libvirt_pci_passthrough_hotplug
+    variants:
+        - NIC:
```
Comment (balamuruhans) [inline libvirt/tests/cfg/libvirt_pci_passthrough_hotplug.cfg]: @Testcases can be organized in this way for better implementation,
```
variants:
    - NIC:
    - Infiniband:
variants:
    - operations:
        variants:
            - hotplug:
            - hotunplug:
            - attach:
            - detach:
            - reattach:
variants:
    - lifecycle:
        variants;
            - suspend:
            - resume:
            - reboot:

```

**Comment by chunfuwen** [comment]
For two files:libvirt_pci_passthrough_hotplug.cfg,libvirt_pci_passthrough_hotplug.py
libvirt_ prefix can be removed as per our naming convention.

File: libvirt/tests/src/libvirt_pci_passthrough_hotplug.py [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: 2017 IBM
```
Comment (chunfuwen) [inline libvirt/tests/src/libvirt_pci_passthrough_hotplug.py]: I do observed that there is no copyright declaration in other py files. And copy right is quite complicated issues, I am not confident about this, are you sure it is necessary to be added  here?

File: libvirt/tests/src/libvirt_pci_passthrough_hotplug.py [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+# Author: Prudhvi Miryala<mprudhvi@linux.vnet.ibm.com>
+# Host device hotplug test
+
+
+import os
+import logging
+import aexpect
+import time
+import re
```
Comment (chunfuwen) [inline libvirt/tests/src/libvirt_pci_passthrough_hotplug.py]: better add one blank line here to separate distinct modules group imports

File: libvirt/tests/src/libvirt_pci_passthrough_hotplug.py (hunk 6 above)
Comment (prudmirya) [reply libvirt/tests/src/libvirt_pci_passthrough_hotplug.py]: yes

File: libvirt/tests/src/libvirt_pci_passthrough_hotplug.py (hunk 2 above)
Comment (prudmirya) [reply libvirt/tests/src/libvirt_pci_passthrough_hotplug.py]: ok

**Comment by balamuruhans** [comment]
can we close this PR as it is duplicate of https://github.com/autotest/tp-libvirt/pull/1411

**Pull request closed without merging**